src/backend/pipeline/cont_adhoc.c

```diff
@@ -0,0 +1,1000 @@
+#include "utils/snapmgr.h"
+#include "utils/syscache.h"
+#include "utils/builtins.h"
+#include <unistd.h>
+
+#include <sys/time.h>
+#include <sys/resource.h>
+
+/* 
+ * This module performs work analogous to the bg workers and combiners
+ * (and a view select), but does it in memory using tuple stores.
+ *
+ * The flow of data is roughly like so:
+ *
+ * WorkerTupleBuffer -> Worker -> Combiner -> View -> Frontend
+ */
+
+typedef struct {
+	TupleBufferBatchReader  *reader;
+	Oid                     view_id;
+	ContinuousView          *view;
+	DestReceiver            *dest;
+	QueryDesc               *query_desc;
+	CQStatEntry             stats;
+	TimestampTz 			last_processed;
+	uint32 					num_processed;
+
+} AdhocWorkerState;
+
+typedef struct {
+	QueryDesc               *query_desc;
+
+	TupleDesc               tup_desc;
+	TupleTableSlot          *slot;
+
+	Tuplestorestate *batch;      /* to hold pre combine rows */
+	TupleHashTable existing; 	 /* only needed if is_agg is true */
+	Tuplestorestate *result; 	 /* holds the combined result */
+
+	DestReceiver            *dest;
+
+	bool is_agg;
+
+} AdhocCombinerState;
+
+typedef struct {
+
+	Tuplestorestate *batch;
+	QueryDesc *query_desc;
+	DestReceiver *dest;
+	TupleTableSlot          *slot;
+
+} AdhocViewState;
+
+static TupleHashTable
+create_agg_hash(int num_cols, Oid *grp_ops, AttrNumber *group_atts)
+{
+	FmgrInfo *eq_funcs = 0;
+	FmgrInfo *hash_funcs = 0;
+	MemoryContext hash_tmp_cxt;
+
+	execTuplesHashPrepare(num_cols, grp_ops,
+						  &eq_funcs, &hash_funcs);
+
+	/* this tmp cxt is required by hash table, it will reset it internally */
+
+	hash_tmp_cxt = AllocSetContextCreate(CurrentMemoryContext,
+			"AdhocCombinerQueryHashTableTempCxt",
+			ALLOCSET_DEFAULT_MINSIZE,
+			ALLOCSET_DEFAULT_INITSIZE,
+			ALLOCSET_DEFAULT_MAXSIZE);
+
+	return BuildTupleHashTable(num_cols, group_atts,
+							   eq_funcs, hash_funcs, 1000,
+							   sizeof(HeapTupleEntryData),
+							   CurrentMemoryContext, hash_tmp_cxt);
+}
+
+static bool
+should_read_fn(TupleBufferReader *reader, TupleBufferSlot *slot)
+{
+	return slot->tuple->group_hash == reader->proc->group_id;
+}
+
+static void
+init_cont_query_batch_context()
+{
+	ContQueryBatchContext = AllocSetContextCreate(CurrentMemoryContext,
+			"ContQueryBatchContext",
+			ALLOCSET_DEFAULT_MINSIZE,
+			ALLOCSET_DEFAULT_INITSIZE,
+			ALLOCSET_DEFAULT_MAXSIZE);
+}
+
+/* util func to set reader on stream scan nodes */ 
+static void
+set_reader(PlanState *planstate, TupleBufferBatchReader *reader)
+{
+	if (planstate == NULL)
+		return;
+
+	if (IsA(planstate, StreamScanState))
+	{
+		StreamScanState *scan = (StreamScanState *) planstate;
+		scan->reader = reader;
+		return;
+	}
+	else if (IsA(planstate, SubqueryScanState))
+	{
+		set_reader(((SubqueryScanState *) planstate)->subplan, reader);
+		return;
+	}
+
+	set_reader(planstate->lefttree, reader);
+	set_reader(planstate->righttree, reader);
+}
+
+static EState *
+create_estate(QueryDesc *query_desc)
```

Comment:
Same for these functions, not sure what the best place to put them would be though.